experimental/unitsconversion/src/converter.rs

```diff
@@ -0,0 +1,345 @@
+        debug_assert!(second_insert_result.is_ok());
+
+        if first_insert_result.is_err() || second_insert_result.is_err() {
+            return Convertibility::NotConvertible;
+        }
+
+        let (sums_are_zeros, subtractions_are_zeros) = map.iter().fold(
+            (true, true),
+            |(sums, subs), (_, determine_convertibility)| {
+                (
+                    sums && determine_convertibility.sum == 0,
+                    subs && determine_convertibility.difference == 0,
+                )
+            },
+        );
+
+        if subtractions_are_zeros {
+            Convertibility::Convertible
+        } else if sums_are_zeros {
+            Convertibility::Reciprocal
+        } else {
+            Convertibility::NotConvertible
+        }
+    }
+
+    fn apply_si_prefix(si_prefix: &SiPrefix, ratio: &mut Ratio<BigInt>) {
+        match si_prefix.base {
+            Base::Decimal => {
+                *ratio *= Ratio::<BigInt>::from_integer(10.into()).pow(si_prefix.power as i32);
+            }
+            Base::Binary => {
+                *ratio *= Ratio::<BigInt>::from_integer(2.into()).pow(si_prefix.power as i32);
+            }
+        }
+    }
+
+    fn compute_conversion_term(
+        &self,
+        unit_item: &MeasureUnitItem,
+        sign: i8,
+    ) -> Option<Ratio<BigInt>> {
+        let conversion_info = self.payload.convert_infos.get(unit_item.unit_id as usize)?;
+
+        let mut conversion_info_factor = Self::extract_ratio_from_unaligned(
+            &conversion_info.factor_sign,
+            conversion_info.factor_num(),
+            conversion_info.factor_den(),
+        );
+
+        Self::apply_si_prefix(&unit_item.si_prefix, &mut conversion_info_factor);
+        // Apply power.
+        conversion_info_factor = conversion_info_factor.pow((unit_item.power * sign) as i32);
+        Some(conversion_info_factor)
+    }
+
+    fn extract_ratio_from_unaligned(
+        sign_ule: &SignULE,
+        num_ule: &ZeroSlice<u8>,
+        den_ule: &ZeroSlice<u8>,
+    ) -> Ratio<BigInt> {
+        let sign = match Sign::from_unaligned(*sign_ule) {
+            Sign::Positive => num_bigint::Sign::Plus,
+            Sign::Negative => num_bigint::Sign::Minus,
+        };
+
+        Ratio::<BigInt>::new(
+            BigInt::from_bytes_le(sign, num_ule.as_ule_slice()),
+            BigInt::from_bytes_le(num_bigint::Sign::Plus, den_ule.as_ule_slice()),
+        )
+    }
+
+    fn compute_offset(
+        &self,
+        input_unit: &MeasureUnit,
+        output_unit: &MeasureUnit,
+    ) -> Option<Ratio<BigInt>> {
+        // In order to have an offset, the input and output units should be simple.
+        // This means, the input and output units should have only one unit item with power 1 and si prefix 0.
+        // For example:
+        //           1 - `meter` and `foot` are simple units.
+        //           2 - `meter-per-second` and `foot-per-second` are not simple units.
```

Comment:
Issue: I think you should instead verify that the units all have zero-valued offsets. The only case where this _is_ valid is if both units are simple and if neither unit is in the denominator.

Reply:
I did not get it, I added to the discussion meeting

Reply:
Let's make up some units to think about this

- excess-person-foot = foot - 3ft
- excess-person-meter = meter - 1m

Those aren't convertible.

- @younies - `meter-per-celsius` is convertible to `meter-per-fahrenheit`, and `liter-celsius` to `liter-fahrenheit`
- @Manishearth - It should be meter per Kelvin and Rankine. Or "degrees fahrenheit" and "degrees celsius"
- @Manishearth - We could just define in our data model that the offset has the behavior as @younies has proposed and we clearly document that. There are only these two temperature units with this problem.
- @robertbastian - Or you could make Celsius and Fahrenheit be a different type, not the same MeasureUnit type.
- @sffc - Also note that CLDR is adding the Beaufort unit.
- @robertbastian - Beaufort doesn't make sense because it is buckets, not a linear unit.
- @robertbastian - We could reduce code size by making a converter type that doesn't have offset. So then there could be a separate TemperatureConverter.
- @sffc - Or in general, we need a higher-level converter type that supports mixed units. So we could make it a Converter enum between the LinearConverter, OffsetConverter, and MixedConverter. And we just define that when you put `fahrenheit` and `celsius` in a LinearConverter, they behave as if they were `rankine` and `kelvin`. Only `Converter` is public; the others are internal things which don't need to be public.
- @younies - But ICU4C supports `square-celsius`
- @Manishearth - That is `square-kelvin`
- @robertbastian - We need to define where the offset gets dropped.

Conclusion:

- Make this PR be a `LinearConverter` without offset
- Make the `OffsetConverter` and `enum Converter { Linear, Offset }` in the next PR

LGTM: @younies @sffc @Manishearth @robertbastian